bitwarden_access_manager.py

```diff
@@ -0,0 +1,672 @@
+        else:
+            raise Exception("Invalid diff operation.")
+
+
+T = TypeVar("T", bound="Diffable")
+
+
+class Diffable(Protocol):
+    def __eq__(self: T, other: Any) -> bool:
+        ...
+
+    def __lt__(self: T, other: T) -> bool:
+        ...
+
+    def get_id(self: T) -> int | str:
+        ...
+
+    def format_toml(self: T) -> str:
+        ...
+
+
+@dataclass(frozen=True)
+class DiffEntry(Generic[T]):
+    actual: T
+    target: T
+
+
+@dataclass(frozen=True)
+class Diff(Generic[T]):
+    to_add: List[T]
+    to_remove: List[T]
+    to_change: List[DiffEntry[T]]
+
+    @staticmethod
+    def new(target: Set[T], actual: Set[T]) -> Diff[T]:
+        # A very basic diff is to just look at everything that needs to be added
+        # and removed, without deeper inspection.
+        to_add = sorted(target - actual)
+        to_remove = sorted(actual - target)
+
+        # However, that produces a very rough diff. If we change e.g. the
+        # description of a group, that would show up as deleting one group and
+        # adding back another which is almost the same, except with a different
+        # description. So to improve on this a bit, if entries have ids, and
+        # the same id needs to be both added and removed, then instead we record
+        # that as a "change".
+        to_add_by_id = {x.get_id(): x for x in to_add}
+        to_remove_by_id = {x.get_id(): x for x in to_remove}
+        to_change = [
+            DiffEntry(
+                actual=to_remove_by_id[id_],
+                target=to_add_by_id[id_],
+            )
+            for id_ in sorted(to_add_by_id.keys() & to_remove_by_id.keys())
+        ]
+
+        # Now that we turned some add/remove pairs into a "change", we should no
+        # longer count those as added/removed.
+        for change in to_change:
+            to_add.remove(change.target)
+            to_remove.remove(change.actual)
+
+        return Diff(
+            to_add=to_add,
+            to_remove=to_remove,
+            to_change=to_change,
+        )
+
+    def print_diff(
+        self,
+        header_to_add: str,
+        header_to_remove: str,
+        header_to_change: str,
+    ) -> None:
+        if len(self.to_add) > 0:
+            print(header_to_add)
+            for entry in self.to_add:
+                print()
+                print_indented(entry.format_toml())
+
+            print()
+
+        if len(self.to_remove) > 0:
+            print(header_to_remove)
+            for entry in self.to_remove:
+                print()
+                print_indented(entry.format_toml())
+
+            print()
+
+        if len(self.to_change) > 0:
+            print(header_to_change)
+            for change in self.to_change:
+                print()
+                print_simple_diff(
+                    actual=change.actual.format_toml(),
+                    target=change.target.format_toml(),
+                )
+
+            print()
```

Comment:
Also here, I did not review `Diffable` at all, I assume it’s the same as in the other script.

Reply:
Yes this is also a copy from Github script.